streaming/base/util.py

```diff
@@ -209,6 +215,179 @@ def get_import_exception_message(package_name: str, extra_deps: str) -> str:
+            shards += obj['shards']
+
+        # Save merged index locally
+        obj = {
+            'version': 2,
+            'shards': shards,
+        }
+        merged_index_path = os.path.join(temp_root, index_basename)
+        with open(merged_index_path, 'w') as outfile:
+            json.dump(obj, outfile)
+
+        # Move merged index from temp path to local part in out
+        # Upload merged index to remote if out has remote part
+        shutil.move(merged_index_path, cu.local)
+        if cu.remote is not None:
+            cu.upload_file(index_basename)
+
+        # Clean up
+        if not keep_local:
+            shutil.rmtree(cu.local, ignore_errors=True)
+
+
+def _merge_index_from_root(out: Union[str, Tuple[str, str]], keep_local: bool = True) -> None:
+    """Merge index.json given the root of MDS dataset. Write merged index to the root folder.
+
+    Args:
+        out (Union[str, Tuple[str,str]]): folder that contain MDS partitions.
+            :A local directory, merge index happens locally
+            :A remote directory, download all the sub-directories index.json in a temporary sub-directories, merge locally, and then upload it to out location
+            :A (local_dir, remote_dir), check if sub-directories index.json file present locally
+                If yes, then merge locally and upload to remote_dir .
+                If not, download all the sub-directories index.json from remote to local , merge locally, and upload to remote_dir .
+        keep_local (bool): Keep local copy of the merged index file. Defaults to ``True``
+    """
+    from streaming.base.storage.upload import CloudUploader
+
+    def not_merged_index(index_file_path: str, out: str):
+        """Check if index_file_path is the merged index at folder out.
+
+        Args:
+            index_file_path (str): the path to index.json file
+            out (str): remote or local url of a folder
+        Return:
+            (bool): no if index.json sits in out instead of in the subfolders of out
+        """
+        prefix = str(urllib.parse.urlparse(out).path)
+        return os.path.dirname(index_file_path).strip('/') != prefix.strip('/')
+
+    if not out:
+        logger.warning('No MDS dataset folder specified, no index merged')
+        return
+
+    cu = CloudUploader.get(out, exist_ok=True, keep_local=True)
+
+    local_index_files = []
+    cl = CloudUploader.get(cu.local, exist_ok=True, keep_local=True)
+    for file in cl.list_objects():
+        if file.endswith('.json') and not_merged_index(file, cu.local):
+            local_index_files.append(file)
+
+    if cu.remote:
+        obj = urllib.parse.urlparse(cu.remote)
+        remote_index_files = []
+        for file in cu.list_objects():
+            if file.endswith(get_index_basename()) and not_merged_index(file, cu.remote):
+                remote_index_files.append(obj.scheme + '://' + os.path.join(obj.netloc, file))
```

Comment:
For UC Volume and DBFS, it's `:/`.

Reply:
changed. 
btw, UC Volume does not support list_objects yet. We should keep an eye on that when they make the endpoint available.